kyo-core/shared/src/main/scala/kyo/Fiber.scala

```diff
@@ -1,746 +1,527 @@
+                new IOPromise[Any, A < S]:
+                    override def interrupt(error: Result.Error[Any]): Boolean = false
+
+            private[kyo] def fromIOPromise[A, S](p: IOPromise[?, A < S]): Unsafe[A, S] = p.asInstanceOf[Unsafe[A, S]]
+
+            extension [A, S](self: Unsafe[A, S])(using NotGiven[S <:< Abort[Any]])
+                private def lower: IOPromise[Nothing, A < S] =
+                    self.asInstanceOf[IOPromise[Nothing, A < S]]
+                def complete(v: Result[Nothing, A < S])(using AllowUnsafe): Boolean =
+                    self.lower.complete(v)
+                def completeDiscard(v: Result[Nothing, A < S])(using AllowUnsafe): Unit =
+                    discard(self.lower.complete(v))
+                def become(other: Fiber[A, S])(using AllowUnsafe): Boolean =
+                    self.lower.become(Fiber.lower(other))
+                def becomeDiscard(other: Fiber[A, S])(using AllowUnsafe): Unit =
+                    discard(lower.become(Fiber.lower(other)))
+                def waiters()(using AllowUnsafe): Int =
+                    self.lower.waiters()
+                def poll()(using AllowUnsafe): Maybe[Result[Nothing, A < S]] =
+                    self.lower.poll()
+                def safe: Promise[A, S] = self
+            end extension
+
+            extension [E, A, S](self: Unsafe[A, Abort[E] & S])
+                private def lower: IOPromise[E, A < S] =
+                    self.asInstanceOf[IOPromise[E, A < S]]
+                def complete(v: Result[E, A < S])(using AllowUnsafe): Boolean =
+                    self.lower.complete(v)
+                def completeDiscard(v: Result[E, A < S])(using AllowUnsafe): Unit =
+                    discard(self.lower.complete(v))
+                def become(other: Fiber[A, Abort[E] & S])(using AllowUnsafe): Boolean =
+                    self.lower.become(Fiber.lower(other))
+                def becomeDiscard(other: Fiber[A, Abort[E] & S])(using AllowUnsafe): Unit =
+                    discard(lower.become(Fiber.lower(other)))
+                def waiters()(using AllowUnsafe): Int =
+                    self.lower.waiters()
+                def poll()(using AllowUnsafe): Maybe[Result[E, A < S]] =
+                    self.lower.poll()
+                def safe: Promise[A, Abort[E] & S] = self
             end extension
         end Unsafe
     end Promise
 
-    private inline def foreach[A](l: Iterable[A])(inline f: (Int, A) => Unit): Unit =
-        l match
-            case l: IndexedSeq[A] =>
-                val s = l.size
-                @tailrec def loop(i: Int): Unit =
-                    if i < s then
-                        f(i, l(i))
-                        loop(i + 1)
-                loop(0)
-            case _ =>
-                val it = l.iterator
-                @tailrec def loop(i: Int): Unit =
-                    if it.hasNext then
-                        f(i, it.next())
+    private[kyo] object internal:
```

Comment:
I've moved these impls to facilitate reading of the source file since they're internal